include/seqan3/std/charconv_detail.hpp

```diff
@@ -758,40 +758,46 @@ inline std::from_chars_result from_chars_floating_point(char const * first,
     // correct parsing. This will also automatically lead to an error if the
     // original sequence did contain a 0x prefix and thus reflect the correct
     // requirements of std::from_chars.
+    ptrdiff_type offset{0};
     if (fmt == std::chars_format::hex)
     {
-        hex_buffer[0] = '0';
-        hex_buffer[1] = 'x';
-        for (unsigned i = 0; i < std::min(buffer_size - 2, last - first); ++i)
-            hex_buffer[i+2] = first[i];
-
-        start = &hex_buffer[0];
-        end = &hex_buffer[0] + sizeof(hex_buffer);
-    }
-    else
-    {
-        start = const_cast<char *>(first);
-        end = const_cast<char *>(last);
+        buffer[0] = '0';
+        buffer[1] = 'x';
+        offset = 2;
     }
 
+    // In contrast to std::from_chars, std::strto[f/d/ld] does not treat the second
+    // parameter (str_end) as "end of the sequence to parse" but merely as an out
+    // parameter to indicate where the parsing ended. Therefore, a buffer is
+    // needed to represent the truncated sequence and ensure correct from_chars
+    // functionality.
+    ptrdiff_type str_length{std::min(ptrdiff_type{buffer_size - offset}, last - first)};
+
+    for (ptrdiff_type i = 0; i < str_length; ++i)
+        buffer[i+offset] = first[i];
+
+    buffer[str_length] = '\0';
+
+    char * end;
+
     if constexpr (std::Same<std::remove_reference_t<value_type>, float>)
     {
-        tmp = strtof(start, &end);
+        tmp = strtof(&buffer[0], &end);
     }
     if constexpr (std::Same<std::remove_reference_t<value_type>, double>)
     {
-        tmp = strtod(start, &end);
+        tmp = strtod(&buffer[0], &end);
     }
     if constexpr (std::Same<std::remove_reference_t<value_type>, long double>)
     {
-        tmp = strtold(start, &end);
+        tmp = strtold(&buffer[0], &end);
     }
 
     if (errno == ERANGE)
     {
         return {last, std::errc::result_out_of_range};
     }
-    else if (tmp == 0 && end == first)
+    else if (tmp == 0 && end == &buffer[0])
     {
         return {last, std::errc::invalid_argument};
     }
```

Comment:
As discussed I would recommend only copying to the buffer when the input is not alread 0-terminated.